### PR TITLE
Fix problem with group.update api

### PR DESCRIPTION
**Pull request by samuelsov**
Overview
----------------------------------------
Following PR #26200, there are 3 things to fix/improve in `CRM_Contact_BAO_Group::update`:
- in the api, the expectation is that if you omit a param then it will not be changed, however the logic is that if we omit $params['parents'] it will be deleted
- there is nothing preventing adding itself as a parent which causes a recursive loop in `CRM_Contact_BAO_GroupNesting::getParentGroupIds` 
- we don't need to update group nesting cache for every update that doesn't touch parents

### CODE REVIEW

File: CRM/Contact/BAO/Group.php [hunk 1]
```diff
@@ -447,28 +447,27 @@ public static function create(&$params) {
         $params['parents'] = [$domainGroupID];
       }
 
-      // first deal with removed parents
-      if (array_key_exists('parents', $params) && !empty($currentParentGroupIDs)) {
-        foreach ($currentParentGroupIDs as $parentGroupId) {
-          // no more parents or not in the new list, let's remove
-          if (empty($params['parents']) || !in_array($parentGroupId, $params['parents'])) {
-            CRM_Contact_BAO_GroupNesting::remove($parentGroupId, $params['id']);
+      if (!CRM_Utils_System::isNull($params['parents'])) {
```
Comment (samuelsov) [inline CRM/Contact/BAO/Group.php:450]: Hmm, `CRM_Utils_System::isNull` is somewhat confusing. Empty array is considered Null so this doesn't work.

**Comment by colemanw** [comment]
@samuelsov probably just delete line 339?

**Comment by samuelsov** [comment]
> @samuelsov probably just delete line 339?

It is converted to an array a bit later in the code so I moved the check to the very beginning of the method.

File: CRM/Contact/BAO/Group.php [hunk 2]
```diff
@@ -384,6 +388,10 @@ public static function create(&$params) {
 
     if (!CRM_Utils_System::isNull($params['parents'])) {
       $params['parents'] = CRM_Utils_Array::convertCheckboxFormatToArray((array) $params['parents']);
```
Comment (colemanw) [inline CRM/Contact/BAO/Group.php:390]: Refactoring note: not a blocker for this PR, but this function really shouldn't be called from the BAO as it's only needed on the quickform.

File: CRM/Contact/BAO/Group.php [hunk 3]
```diff
@@ -384,6 +388,10 @@ public static function create(&$params) {
 
     if (!CRM_Utils_System::isNull($params['parents'])) {
```
Comment (colemanw) [inline CRM/Contact/BAO/Group.php:389]:
```suggestion
    if ($parentsParamProvided) {
```

Comment (samuelsov) [reply CRM/Contact/BAO/Group.php]: Ok, done.

**Comment by colemanw** [comment]
@samuelsov are you able to rebase this into a single commit?

**Comment by samuelsov** [comment]
@colemanw ok, yeah, done (I think). I though you could do it when merging like in gitlab.

**Comment by colemanw** [comment]
That would be nice! But our workflow currently requires a merge-commit.

@samuelsov I think the only thing missing here is a unit test. Are you up for that?

**Comment by samuelsov** [comment]
@colemanw yes, hopefully tomorrow :)

**Comment by colemanw** [comment]
That would be great thanks @samuelsov

**Comment by samuelsov** [comment]
@colemanw seems good to me, the tests detected an error that I fixed. Can you review ?

File: CRM/Contact/BAO/Group.php [hunk 4]
```diff
@@ -448,7 +456,7 @@ public static function create(&$params) {
       }
```
Comment (colemanw) [inline CRM/Contact/BAO/Group.php:456]: This block (see above) will actually force in a `parents` param, so we either need to alter `$parentsParamProvided` here or else modify the check below for refreshing the cache.

Comment (samuelsov) [reply CRM/Contact/BAO/Group.php:456]: Ok, any preference ? we could also go back to a safer approach where the cache is updated no matter what.

File: CRM/Contact/BAO/Group.php [hunk 5]
```diff
@@ -466,9 +474,10 @@ public static function create(&$params) {
         }
       }
 
-      // this is always required, since we don't know when a
-      // parent group is removed
-      CRM_Contact_BAO_GroupNestingCache::update();
+      // refresh cache if parents param was provided
+      if ($parentsParamProvided) {
```
Comment (colemanw) [inline CRM/Contact/BAO/Group.php:478]: How about this?
```suggestion
      if ($parentsParamProvided || !empty($params['parents'])) {
```

**Pull request merged**